sway-core/src/asm_generation/expression/contract_call.rs

```diff
@@ -34,54 +33,111 @@ pub(crate) fn convert_contract_call_to_asm(
         owning_span: Some(span.clone()),
     });
 
-    // evaluate the user provided argument to the contract
-    asm_buf.append(&mut check!(
-        convert_expression_to_asm(
-            user_argument,
-            namespace,
-            &user_argument_register,
-            register_sequencer
-        ),
-        vec![],
-        warnings,
-        errors
-    ));
-
-    // evaluate the gas to forward to the contract
-    asm_buf.append(&mut check!(
-        convert_expression_to_asm(cgas, namespace, &gas_to_forward, register_sequencer),
-        vec![],
-        warnings,
-        errors
-    ));
-
-    // evaluate the balance to forward to the contract
-    asm_buf.append(&mut check!(
-        convert_expression_to_asm(bal, namespace, &bal_register, register_sequencer),
-        vec![],
-        warnings,
-        errors
-    ));
-
-    // evaluate the coin color expression to forward to the contract
-    asm_buf.append(&mut check!(
-        convert_expression_to_asm(
-            // investigation: changing this value also results in a different color
-            coin_color,
-            namespace,
-            &coin_color_register,
-            register_sequencer
-        ),
-        vec![],
-        warnings,
-        errors
-    ));
+    // Bundle the arguments into a struct if needed. If we only have a single argument, there is no
+    // need to bundle.
+    let bundled_arguments = match arguments.len() {
+        0 => None,
+        1 => Some(arguments[0].1.clone()),
+        _ => {
+            // create a struct expression that bundles the arguments in order
+            let mut typed_fields_buf = vec![];
+            for (name, arg) in arguments {
+                typed_fields_buf.push(TypedStructExpressionField {
+                    value: arg.clone(),
+                    name: name.clone(),
+                });
+            }
+            Some(TypedExpression {
+                expression: TypedExpressionVariant::StructExpression {
+                    struct_name: Ident::new_with_override("bundled_arguments", span.clone()),
+                    fields: typed_fields_buf,
+                },
+                return_type: 0,
+                is_constant: IsConstant::No,
+                span: span.clone(),
+            })
+        }
+    };
+
+    // evaluate the bundle of arguments
+    if let Some(bundled_arguments) = &bundled_arguments {
+        asm_buf.append(&mut check!(
+            convert_expression_to_asm(
+                bundled_arguments,
+                namespace,
+                &bundled_arguments_register,
+                register_sequencer
+            ),
+            vec![],
+            warnings,
+            errors
+        ));
+    }
+
+    // evaluate the gas to forward to the contract. If no user-specified gas parameter is found,
+    // simply load $cgas.
```

Comment:
BTW, I think loading `$zero` would also be valid here as 0 is shorthand for forwarding all remaining gas. CC @adlerjohn for a correctness check on that

Reply:
I merged for now while I can. Will follow up with John on Monday and change if necessary :)